src/parallel/parallel.cpp

```diff
@@ -123,32 +123,192 @@ ThreadManager::get_segment_index(size_t index)
+    }
+    if(key_exists) {
+        return false;
+    }
+    ThreadManager manager = this->get_thread_lock_manager();
+
+    size_t offset = 0;
+    ParallelBucket entry_to_swap = {.key = key, .value = value, .hashcode = hashcode, .offset = offset};
+
+    auto [next_index, found] = this->find_next_index_lock(manager, home, key, offset);
+
+    if (found) {
+        manager.release_all_locks();
+        return false;
+    }
+
+    locked_insert(entry_to_swap, next_index);
+    manager.release_all_locks();
+    return true;
 }
 
 void
 ParallelRobinHoodHashTable::insert_or_update(KeyType key, ValueType value)
 {
-    // TODO
+    ThreadManager manager = this->get_thread_lock_manager();
+    HashCodeType hashcode = this->hash(key);
+    size_t home = this->get_home(hashcode, this->capacity_);
+    size_t idx = home;
+
+    KeyType prev_key = key;
+    ValueType prev_val = value;
+    //bool is_inserted, is_found, found;
+
+    for(size_t distance = 0; this->buckets_[home].offset >= distance; ++distance, ++idx){
+        while(this->buckets_[idx].key == key) {
+            KeyType prev_key_copy = prev_key;
+            ValueType prev_val_copy = prev_val;
+            auto [prev_key, prev_val] = this->compare_and_set_key_val(home, prev_key_copy, this->buckets_[idx].key, prev_val_copy);
+            if (prev_key == key) {
+                return;
+            }
+        }
+    }
+    auto [is_inserted, is_found] = this->distance_zero_insert(key, value, home);
+    if(is_inserted) {
+        return;
+    }
+    if(is_found) { //not sure about these?
+        return;
+    }
+    size_t offset = 0;
+    ParallelBucket entry_to_insert = {.key = key, .value = value, .hashcode = hashcode, .offset = offset}; //offset is arbitrary
+    auto [next_index, found] = this->find_next_index_lock(manager, home, key, offset);
+
+    if(found){
+        this->buckets_[next_index].value = value;
+        manager.release_all_locks();
+        return;
+    }
+
+    this->locked_insert(entry_to_insert, next_index);
+    manager.release_all_locks();
+    return;
 }
 
 bool
 ParallelRobinHoodHashTable::remove(KeyType key, ValueType value)
 {
-    // TODO
+    HashCodeType hashcode = this->hash(key);
+    size_t home = this->get_home(hashcode, this->capacity_);
+    ThreadManager manager = this->get_thread_lock_manager();
+    ParallelBucket entry_to_insert = {.key = key, .value = value, .hashcode = hashcode, .offset = 0}; //offset is arbitrary
+
+    auto[index, found] = this->find_next_index_lock(manager, home, key, entry_to_insert.offset);
+
+    if(!found) {
+        manager.release_all_locks();
+        return false;
+    }
+
+    size_t next_index = index++;
+    size_t curr_index = index;
+    
+    manager.lock(curr_index);
+    while(this->buckets_[next_index].offset > 0 && !entry_to_insert.is_empty()) {
+        manager.lock(next_index); 
+        ParallelBucket &entry_to_move = this->do_atomic_swap(entry_to_insert, next_index); //na this is fucked up 
```

Comment:
```suggestion
        ParallelBucket entry_to_move = this->do_atomic_swap(locked_entry, next_index);
```
We shouldn't be swapping in the entry to delete.